### PR TITLE
waybar: include the systemd user unit

### PR DESCRIPTION
**Pull request by r-vdp**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by khaneliman** [comment]
Looks like a duplicate of https://github.com/NixOS/nixpkgs/pull/242728

**Comment by r-vdp** [comment]
Right. I included the change to no longer define the service in the module.

Compared to the other PR, this one:
1. does not include an unneeded `systemd` input to the nix expression
2. does not set unneeded attributes on the service (the `wantedBy` on `graphical-session.target) is already present in the unit included in the package)

I don't really mind either way, but unless anyone has comments on this PR, we can just merge it instead of fixing up the other one that is a year old.

**Comment by Scrumplex** [comment]
Result of `nixpkgs-review pr 340874` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>disko</li>
    <li>waybar</li>
  </ul>
</details>

File: nixos/modules/programs/wayland/waybar.nix [hunk 1]
```diff
@@ -11,12 +16,6 @@ in
 
   config = lib.mkIf cfg.enable {
     environment.systemPackages = [ cfg.package ];
-    systemd.user.services.waybar = {
-      description = "Waybar as systemd service";
-      wantedBy = [ "graphical-session.target" ];
-      partOf = [ "graphical-session.target" ];
-      script = "${cfg.package}/bin/waybar";
-    };
   };
```
Comment (Scrumplex) [inline nixos/modules/programs/wayland/waybar.nix:19]: We should also add `cfg.package` to `systemd.packages`

**Review by Scrumplex — CHANGES_REQUESTED** [review]
Final unit looks good.

```ini
[Unit]
Description=Highly customizable Wayland bar for Sway and Wlroots based compositors.
Documentation=https://github.com/Alexays/Waybar/wiki/
PartOf=graphical-session.target
After=graphical-session.target
Requisite=graphical-session.target

[Service]
ExecStart=/nix/store/yar3wkz3528wb465v9axk6dj480lpdyq-waybar-0.10.4/bin/waybar
ExecReload=kill -SIGUSR2 $MAINPID
Restart=on-failure

[Install]
WantedBy=graphical-session.target

```

File: nixos/modules/programs/wayland/waybar.nix (hunk 1 above)
Comment (r-vdp) [reply nixos/modules/programs/wayland/waybar.nix:19]: Don't packages in systemPackages get included already? I need to double check.

**Comment by r-vdp** [comment]
I think we may need to add an explicit wantedBy actually, since our nix code to generate etc doesn't look at the install section.

I'll push that change after I had lunch.

File: nixos/modules/programs/wayland/waybar.nix (hunk 1 above)
Comment (r-vdp) [reply nixos/modules/programs/wayland/waybar.nix]: Added it.

**Review by Scrumplex — APPROVED** [review]

**Review by GaetanLepage — APPROVED** [review]
Diff LGTM !

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 340874` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>disko</li>
    <li>waybar</li>
  </ul>
</details>

**Comment by JohnRTitor** [comment]
> I think we may need to add an explicit wantedBy actually, since our nix code to generate etc doesn't look at the install section.

Looks like it does automatically add WantedBy.

**Review by JohnRTitor — APPROVED** [review]

**Comment by r-vdp** [comment]
> > I think we may need to add an explicit wantedBy actually, since our nix code to generate etc doesn't look at the install section.
> 
> Looks like it does automatically add WantedBy (without any involvement from the module)
> 
> image
> 
> But it doesn't hurt to override it anyway
> 

@JohnRTitor yeah but the code that assembles the systemd directory in NixOS doesn't consider a unit's install section, it only takes into consideration the dependencies that are present in the module system options. Systemd also doesn't read this section unless you imperatively enable the service, which no one does on NixOS.

So if we don't add the dependency explicitly in the module, the service won't be pulled in.

**Comment by m-bdf** [comment]
> does not include an unneeded `systemd` input to the nix expression

Well now build fails with:

```sh
Run-time dependency systemd found: NO (tried pkgconfig)
meson.build:112:10: ERROR: Dependency "systemd" not found, tried pkgconfig
```

So maybe it wasn't that unneeded after all :p

---

For context, I disable support for all the modules I don't use:

```nix
pkgs.waybar.override {
  cavaSupport = false;
  evdevSupport = false;
  experimentalPatches = false;
  hyprlandSupport = false;
  inputSupport = false;
  jackSupport = false;
  mpdSupport = false;
  mprisSupport = false;
  pulseSupport = false;
  sndioSupport = false;
  swaySupport = false;
  traySupport = false;
  udevSupport = false;
  upowerSupport = false;
};
```

I guess most people don't do this, and that's why the build failure wasn't caught.

**Comment by r-vdp** [comment]
Hmm, weird that the build doesn't fail with those enabled then. I guess we should add the systemd build-dep then, indeed. Sorry about that.

**Comment by Scrumplex** [comment]
See #345107

**Comment by acid-bong** [comment]
But does it enable the service? I tried to do the same thing with user services from my config and it didn't seem so. Maybe `systemd.user.services.waybar.enable = true` is needed?

**Comment by r-vdp** [comment]
> But does it enable the service? I tried to do the same thing with user services from my config and it didn't seem so. Maybe `systemd.user.services.waybar.enable = true` is needed?

It should get started by graphical-session.target, and that target is supposed to be started by your compositor.

**Pull request merged**